src/reanimated2/animation/transformationMatrix/matrixUtils.tsx

```diff
@@ -0,0 +1,433 @@
+        row.length === 4 &&
+        row.every((element) => typeof element === 'number')
+    )
+  );
+}
+
+export function flatten(matrix: AffiniteMatrix): AffiniteMatrixFlat {
+  'worklet';
+  return matrix.flat() as AffiniteMatrixFlat;
+}
+
+export function unflatten(matrix: AffiniteMatrixFlat): AffiniteMatrix {
+  'worklet';
+
+  return [
+    [matrix[0], matrix[1], matrix[2], matrix[3]],
+    [matrix[4], matrix[5], matrix[6], matrix[7]],
+    [matrix[8], matrix[9], matrix[10], matrix[11]],
+    [matrix[12], matrix[13], matrix[14], matrix[15]],
+  ] as AffiniteMatrix;
+}
+
+function maybeFlattenMatrix(
+  matrix: AffiniteMatrix | AffiniteMatrixFlat
+): AffiniteMatrixFlat {
+  'worklet';
+
+  return isAffiniteMatrix(matrix) ? flatten(matrix) : matrix;
+}
+
+export function multiplyMatrices(
+  a: AffiniteMatrix,
+  b: AffiniteMatrix
+): AffiniteMatrix {
+  'worklet';
+  return [
+    [
+      a[0][0] * b[0][0] +
+        a[0][1] * b[1][0] +
+        a[0][2] * b[2][0] +
+        a[0][3] * b[3][0],
+
+      a[0][0] * b[0][1] +
+        a[0][1] * b[1][1] +
+        a[0][2] * b[2][1] +
+        a[0][3] * b[3][1],
+
+      a[0][0] * b[0][2] +
+        a[0][1] * b[1][2] +
+        a[0][2] * b[2][2] +
+        a[0][3] * b[3][2],
+
+      a[0][0] * b[0][3] +
+        a[0][1] * b[1][3] +
+        a[0][2] * b[2][3] +
+        a[0][3] * b[3][3],
+    ],
+    [
+      a[1][0] * b[0][0] +
+        a[1][1] * b[1][0] +
+        a[1][2] * b[2][0] +
+        a[1][3] * b[3][0],
+
+      a[1][0] * b[0][1] +
+        a[1][1] * b[1][1] +
+        a[1][2] * b[2][1] +
+        a[1][3] * b[3][1],
+
+      a[1][0] * b[0][2] +
+        a[1][1] * b[1][2] +
+        a[1][2] * b[2][2] +
+        a[1][3] * b[3][2],
+
+      a[1][0] * b[0][3] +
+        a[1][1] * b[1][3] +
+        a[1][2] * b[2][3] +
+        a[1][3] * b[3][3],
+    ],
+    [
+      a[2][0] * b[0][0] +
+        a[2][1] * b[1][0] +
+        a[2][2] * b[2][0] +
+        a[2][3] * b[3][0],
+
+      a[2][0] * b[0][1] +
+        a[2][1] * b[1][1] +
+        a[2][2] * b[2][1] +
+        a[2][3] * b[3][1],
+
+      a[2][0] * b[0][2] +
+        a[2][1] * b[1][2] +
+        a[2][2] * b[2][2] +
+        a[2][3] * b[3][2],
+
+      a[2][0] * b[0][3] +
+        a[2][1] * b[1][3] +
+        a[2][2] * b[2][3] +
+        a[2][3] * b[3][3],
+    ],
+    [
+      a[3][0] * b[0][0] +
+        a[3][1] * b[1][0] +
+        a[3][2] * b[2][0] +
+        a[3][3] * b[3][0],
+
+      a[3][0] * b[0][1] +
+        a[3][1] * b[1][1] +
+        a[3][2] * b[2][1] +
+        a[3][3] * b[3][1],
+
+      a[3][0] * b[0][2] +
+        a[3][1] * b[1][2] +
+        a[3][2] * b[2][2] +
+        a[3][3] * b[3][2],
+
+      a[3][0] * b[0][3] +
+        a[3][1] * b[1][3] +
+        a[3][2] * b[2][3] +
+        a[3][3] * b[3][3],
+    ],
+  ];
+}
+
+function linearOperationOnMatrix<T extends AffiniteMatrixFlat | AffiniteMatrix>(
+  _a: T,
+  func: (x: number) => number
+): T {
+  'worklet';
+
+  const isFlatOnStart = !!isAffiniteMatrixFlat(_a);
```

Comment:
Do we need `!!` here? It looks like `isAffiniteMatrixFlat` always returns a boolean.